features/net/network-socket/WiFiInterface.h

```diff
@@ -43,28 +42,47 @@ class WiFiInterface: public NetworkInterface
 public:
     /** Start the interface
      *
-     *  Attempts to connect to a WiFi network. If passphrase is invalid,
-     *  NSAPI_ERROR_AUTH_ERROR is returned.
+     *  Attempts to connect to a WiFi network.
      *
      *  @param ssid      Name of the network to connect to
      *  @param pass      Security passphrase to connect to the network
      *  @param security  Type of encryption for connection
-     *  @return          0 on success, negative error code on failure
+     *  @return          0 on success, or error code on failure
      */
-    virtual int connect(const char *ssid, const char *pass, nsapi_security_t security = NSAPI_SECURITY_NONE) = 0;
+    virtual nsapi_error_t connect(const char *ssid, const char *pass, nsapi_security_t security = NSAPI_SECURITY_NONE) = 0;
 
     /** Stop the interface
      *
-     *  @return          0 on success, negative error code on failure
+     *  @return          0 on success, or error code on failure
      */
-    virtual int disconnect() = 0;
+    virtual nsapi_error_t disconnect() = 0;
 
     /** Get the local MAC address
      *
      *  @return         Null-terminated representation of the local MAC address
      */
     virtual const char *get_mac_address() = 0;
-};
 
+    /** Get the current WiFi interface state
+     *
+     * @returns         Interface state enum
+     */
+    virtual nsapi_if_state_t get_state() = 0;
+
+    /** Gets the current radio signal strength for active connection
+     *
+     * @return Connection strength in dBm (negative value), or 0 if measurement impossible
+     */
+    virtual int16_t get_rssi() = 0;
+
+    /** Scan for available networks
+     *
+     * @param  ap    Pointer to allocated array to store discovered AP
+     * @param  count Size of allocated @a res array, or 0 to only count available AP
+     * @return       Number of entries in @a, or if @a count was 0 number of available networks, negative on error
+     *               see @a nsapi_error
+     */
+    virtual int scan(wifi_ap_t *res, unsigned count) = 0;
```

Comment:
this might be problematic api for application. Only way is to estimate how many networks there might be available, but still you might get just partial list of ssid's. It would be more usefull and also memory friendly if you could get one by one result, or if this api allow to skip N results...